src/marlin/types.ts

```diff
@@ -0,0 +1,188 @@
+
+// API is imported contracts classes
+export type Api = Comp | GovernorAlpha | Timelock;
+
+export const EventChains = ['marlin', 'marlin-local'] as const;
+
+export type RawEvent = Event;
+
+// TODO: is necessary?
+// export enum EntityKind {
+//   Proposal = 'proposal',
+// }
+
+export enum EventKind {
+  // Comp Events
+  Approval = 'approval',
+  DelegateChanged = 'delegate-changed',
+  DelegateVotesChanged = 'delegate-votes-changed',
+  Transfer = 'transfer',
+  // TODO: GovernorAlpha Events
+  ProposalExecuted = 'proposal-executed',
+  ProposalCreated = 'proposal-created',
+  ProposalCanceled = 'proposal-canceled',
+  ProposalQueued = 'proposal-queued',
+  VoteCast = 'vote-cast',
+  // TODO: Timelock Events
+  CancelTransaction = 'cancel-transaction',
+  ExecuteTransaction = 'execute-transactions',
+  NewAdmin = 'new-admin',
+  NewDelay = 'new-delay',
+  NewPendingAdmin = 'new-pending-admin',
+  QueueTransaction = 'queue-transaction',
+}
+
+interface IEvent {
+  kind: EventKind;
+}
+
+type Address = string;
+type Balance = string;
+
+// Comp Event Interfaces
+export interface IApproval extends IEvent {
+  kind: EventKind.Approval,
+  owner: Address,
+  spender: Address,
+  amount: Balance,
+}
+
+export interface IDelegateChanged extends IEvent {
+  kind: EventKind.DelegateChanged,
+  delegator: Address,
+  fromDelegate: Address,
+  toDelegate: Address,
+}
+
+export interface IDelegateVotesChanged extends IEvent {
+  kind: EventKind.DelegateVotesChanged,
+  delegate: Address,
+  previousBalance: Balance,
+  newBalance: Balance,
+}
+
+export interface ITransfer extends IEvent {
+  kind: EventKind.Transfer,
+  from: Address,
+  to: Address,
+  amount: Balance,
+}
+
+// TODO: GovernorAlpha Event Interfaces
+export interface IProposalCanceled extends IEvent {
+  kind: EventKind.ProposalCanceled,
+  id: number,
+}
+
+export interface IProposalCreated extends IEvent {
+  kind: EventKind.ProposalCreated,
+  id: number,
+  proposer: Address | null,
+  targets: Address[], // in function, encoding string[]
+  values: number[], // BigNumberish[]
+  signatures: Address[], // string[]
+  calldatas: string[], // Arrayish[], TODO: decide on type
+  startBlock: number,
+  endBlock: number,
+  description: string,
+}
+
+export interface IProposalExecuted extends IEvent {
+  kind: EventKind.ProposalExecuted,
+  id: number,
+}
+
+export interface IProposalQueued extends IEvent {
+  kind: EventKind.ProposalQueued,
+  id: number,
+  eta: number,
+}
+
+export interface IVoteCast extends IEvent {
+  kind: EventKind.VoteCast,
+  voter: Address,
+  proposalId: number,
+  support: boolean, 
+  votes: Balance, // TODO: or just number?
+}
+
+
+// TODO: Timelock Event Interfaces
+export interface ICancelTransaction extends IEvent {
+  kind: EventKind.CancelTransaction,
+  txHash: [], // Arrayish, TODO: decide type
+  target: Address,
+  value: Balance,// TODO: or just number?
+  signature: string,
+  data: [], // Arrayish, TODO: decide type
```

Comment:
What sort of data?

Reply:
unclear from the contract, but Arrayish is a union type of String | ArrayLike<number>, so I'm going to assume it a string